### PR TITLE
Add JSON RPC

### PR DESCRIPTION
**Pull request by noxiouz**

### CODE REVIEW

**Comment by noxiouz** [comment]
based on https://github.com/cocaine/cocaine-tools/tree/issue/xx/json-rpc

**Comment by 3Hren** [comment]
Updated, @noxiouz PTAL and, if OK, merge. Compatible with 12.8.

File: cocaine/proxy/jsonrpc.py [hunk 1]
```diff
@@ -24,52 +35,100 @@ def match(self, request):
...
+            'jsonrpc': '2.0',
+            'result': result,
+            'id': payload['id'],
+        }
+        fill_response_in(request, 200, 'OK', json.dumps(body), headers)
+
+    @gen.coroutine
+    def _handle_mute(self, service, method, args, _):
+        yield getattr(service, method)(*args)
+        raise gen.Return(None)
```
Comment (noxiouz) [inline cocaine/proxy/jsonrpc.py]: just ```return```

File: cocaine/proxy/jsonrpc.py [hunk 2]
```diff
@@ -24,52 +35,100 @@ def match(self, request):
...
+
+    @staticmethod
+    def _send_500_error(request, payload, err):
+        headers = httputil.HTTPHeaders({'Content-Type': 'application/json-rpc'})
+        body = {
+            'jsonrpc': '2.0',
+            'error': str(err),
+            'id': payload['id'],
+        }
+        fill_response_in(request, 500, 'Internal Server Error', json.dumps(body), headers)
```
Comment (noxiouz) [inline cocaine/proxy/jsonrpc.py]: newline at the end

File: cocaine/proxy/jsonrpc.py [hunk 3]
```diff
@@ -24,52 +35,100 @@ def match(self, request):
...
-            _, _, rxtree = named_api[service_method]
-            if rxtree != {0: ['value', {}], 1: ['error', {}]}:
-                headers = httputil.HTTPHeaders({'Content-Type': 'application/json-rpc'})
-                body = {'code': -32000, 'message': 'Protocol is not primitive.'}
-                fill_response_in(request, 400, 'Bad JSON-RPC request', json.dumps(body), headers)
+        except Exception as err:
+            JSONRPC._send_500_error(request, payload, err)
+            return
+
+        methods = (data[0] for (_, data) in service.api.iteritems())
```
Comment (noxiouz) [inline cocaine/proxy/jsonrpc.py]:
```python
(data[0] for data in service.api.itervalues())
```

File: cocaine/proxy/jsonrpc.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+        named_api = dict((name, [name, tx, rx]) for (_, (name, tx, rx)) in service.api.iteritems())
+        _, tx_tree, rx_tree = named_api[method]
+
+        try:
+            for match, handle in self._protocols:
+                if match(tx_tree, rx_tree):
+                    result = yield handle(service, method, args, chunks)
+                    break
+            else:
```
Comment (noxiouz) [inline cocaine/proxy/jsonrpc.py]: Yay! For-Else: it fits ideally here

**Comment by 3Hren** [comment]
Fixed.

File: cocaine/proxy/jsonrpc.py [hunk 5]
```diff
@@ -57,13 +57,12 @@ def process(self, request):
             JSONRPC._send_500_error(request, payload, err)
             return
 
-        methods = (data[0] for (_, data) in service.api.iteritems())
-        if method not in methods:
+        api = dict((data[0], data) for data in service.api.itervalues())
+        if method not in api.keys():
```
Comment (noxiouz) [inline cocaine/proxy/jsonrpc.py]: method not in api

**Pull request merged**